packages/next/src/client/components/app-router.tsx

```diff
@@ -283,6 +285,15 @@ function Router({
   const navigate = useNavigate(dispatch)
   useServerActionDispatcher(dispatch)
 
+  const internalAppRouter = useMemo<InternalAppRouterInstance>(() => {
+    return {
+      navigate: (navigationType, href, scroll, linkInstance) => {
+        setLinkForCurrentNavigation(linkInstance)
+        navigate(href, navigationType, scroll)
+      },
+    }
+  }, [navigate])
```

Comment:
We already have too many layers of indirection here and I'm loth to add another one. I cleaned this up a bit in the action queue refactor I've been working on; I'll submit part of that as a PR and then we can rebase this on top.